_posts/2017-03-05-artsy-technology-stack-2017.md

```diff
@@ -0,0 +1,156 @@
+
+We continue to have a [public HAL+JSON API](https://developers.artsy.net) for external developers. This API is in active use for contemporary production services inside Artsy and the [website is open-source](https://github.com/artsy/doppler), too.
+
+
+<center>
+ <img src="/images/tech-2017/languages.svg" style="width:100%;">
+</center>
+
+## Partner-Facing
+
+The vast customer-facing business is powered by a Content Management System (CMS) for gallery and institutional partners. This CMS lets them upload and manage gallery shows, fair booths, create artists, and edit artwork metadata. All CMS components talk to our core API. We also have a number of CMS-like internal applications to manage partners, auctions, art genomes, configuring fairs or performing recurrent billing (we use Stripe for storing and charging credit cards and ACH) with invoicing.
+
+CMS applications are based on stable, mature technologies like [Rails](http://rubyonrails.org), [Bootstrap](http://getbootstrap.com), [Turbolinks](https://github.com/turbolinks/turbolinks) and [CoffeeScript](http://coffeescript.org), and gradually adopts modern client-side technologies like [React](https://facebook.github.io/react) and [Browserify](http://browserify.org). They share a lot of common infrastructure.
+
+We have a generic image-processing service in-house, which uses [Rails](http://rubyonrails.org), [Sidekiq](https://github.com/mperham/sidekiq/), [Redis](https://redis.io), and [RMagick](https://github.com/rmagick/rmagick) with [ImageMagick](http://www.imagemagick.org/script/index.php). It receives image processing requests from many Artsy applications and generates thumbnails, tiles and watermarks images on S3.
+
+## Collector-Facing
+
+Collectors inquire on artworks and engage in conversations with partners. For this purpose we have built a generic messaging system that manages communications between different parties. It receives messages via API or e-mail, finds or creates a conversation based on the recipients and forwards them to the proper addresses in that conversation. Its doesn't assume anything about the contents of the messages, which makes it a generic system for any type of conversation. The conversations surface to our partners via CMS.
+
+## Running Auctions
+
+Artsy's Auctions business started with charity auctions. Charity auctions are simpler to map digitally: they have less bids, are more free form in terms of bid increments, have less artworks for sale, they happen slower and the people running them are less risk-averse as they tend to be one-off annual events instead of regular occurrences. We therefore modeled these Auctions inside the core API. As we started implementing commercial auctions with a real-time component (we call these "Live Auctions") the differences in the scale of the domain made it worth moving the logic around auctions into a separate micro-service.
```

Comment:
I wouldn't necessarily agree that charity auctions are simpler -- they often have complicated logistics, live events, one-off requests, use our kiosks and live displays, offer "experience" lots that aren't art, have less experienced partners (it's not their core competency), and so on. Plus, the fact that they could be the charity's biggest fundraiser for the year makes them high stakes.

I'd maybe say that we began our work with auctions doing the occasional benefit auctions for charities. Most of these auctions are online-only, timed sales. The initial version of our auction systems came together before we began our move to microservices, and so it is baked into our core API. Last year, we launched a live auction integration product to allow users to bid on works at commercial sales at the actual auction house sale rooms. The real-time requirements of this system required a rethinking of how we process our bids.